### PR TITLE
1424: CheckRun fails to evaluate jol#24

### PR DESCRIPTION
**Pull request by erikj79**
CheckRun, which is part of the PullRequestBot, is currently failing to evaluate https://github.com/openjdk/jol/pull/24. The failure happens in PullRequestUtils::containsForeignMerge where we are looking for merges in the PR branch which bring in commits that aren't descendants of the main merge parent. If this test is positive, the bot adds a comment with a warning.

The problem is that GitRepository::mergeBase doesn't differentiate between the git command failing with an error and when it just can't find a merge-base. For this check, a non existent merge-base should qualify for a positive test.

I'm refactoring Repository::mergeBase into two methods, one that returns an Optional, which will be empty when a merge-base can't be found. I've also gone through all callers to identify when the new method should be used to properly tell this situation apart.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [SKARA-1424](https://bugs.openjdk.java.net/browse/SKARA-1424): CheckRun fails to evaluate jol#24

### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/skara pull/1312/head:pull/1312` \
`$ git checkout pull/1312`

Update a local copy of the PR: \
`$ git checkout pull/1312` \
`$ git pull https://git.openjdk.java.net/skara pull/1312/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1312`

View PR using the GUI difftool: \
`$ git pr show -t 1312`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/skara/pull/1312.diff">https://git.openjdk.java.net/skara/pull/1312.diff</a>

</details>

### CODE REVIEW

File: forge/src/main/java/org/openjdk/skara/forge/PullRequestUtils.java [hunk 1]
```diff
@@ -128,11 +128,8 @@ private static Hash findSourceHash(PullRequest pr, Repository localRepo, List<Co
         var sourceHead = fetchMergeSource(pr, localRepo);
 
         // Ensure that the source and the target are related
-        try {
-            localRepo.mergeBase(targetHash(localRepo), sourceHead);
-        } catch (IOException e) {
-            throw new CommitFailure("The target and the source branches do not share common history - cannot merge them.");
-        }
+        localRepo.mergeBaseOptional(targetHash(localRepo), sourceHead)
+                .orElseThrow(() -> new CommitFailure("The target and the source branches do not share common history - cannot merge them."));
```
Comment (erikj79) [inline forge/src/main/java/org/openjdk/skara/forge/PullRequestUtils.java:132]: Here I think a real error should be let through so the whole WorkItem may be retried later.

File: vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java [hunk 2]
```diff
@@ -919,16 +919,28 @@ public void prune(Branch branch, String remote) throws IOException {
     }
 
     @Override
-    public Hash mergeBase(Hash first, Hash second) throws IOException {
+    public Optional<Hash> mergeBaseOptional(Hash first, Hash second) throws IOException {
         try (var p = capture("git", "merge-base", first.hex(), second.hex())) {
-            var res = await(p);
```
Comment (erikj79) [inline vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java:924]: await(p) basically does p.await() and then throws IOException if res.status() is non zero.

**Review by kevinrushforth — APPROVED** [review]
Looks good.

**Pull request closed without merging**